### ISSUE TITLE
object of type '_sre.SRE_Pattern' has no len()

### ISSUE THREAD

**Issue by Bartvds** [ambitioninc/django-regex-field#14]
I'm on Django 1.7 trying to use this field in the django-admin, but while saving it raises:

```
object of type '_sre.SRE_Pattern' has no len()
```

When I look in the stacktrace I see the `MaxLengthValidator` is running on the regular expression instead of the pattern:

```
MaxLengthValidator(BaseValidator):
...
        compare = lambda self, a, b: a > b
        clean = lambda self, x: len(x)
...
▼ Local vars
Variable    Value
x   re.compile('mypattern')
self    <django.core.validators.MaxLengthValidator object at 0xb683a8ec>
```

I didn't do anything special to add it to the admin; just listing it in 'fields' of the ModelAdmin subclass.

**Comment by MikeSandford** [comment]
I am having the same problem in django 1.9.  I removed it from the list display which didn't change anything.

**Comment by MikeSandford** [comment]
So I am thinking about this problem and since I'm not super familiar with how the internals work I can't say for sure if this is right.  But it seems like two things are needed:

1.  Ensuring it's a good regex during validation
2.  Compiling the string into a regex prior to use

On a whim I dove into the django source, guessing that the to_python function was getting called before, but not too far before the run_validators.  So I checked in on django/db/modeds/fields/__init__.py line 600:

```
    def clean(self, value, model_instance):
        """
        Convert the value's type and run validation. Validation errors
        from to_python and validate are propagated. The correct value is
        returned if no error is raised.
        """
        value = self.to_python(value)
        self.validate(value, model_instance)
        self.run_validators(value)
        return value
```

I got super lucky!  I found the exact right spot.  The self.to_python is where the RegexField is turning the string into an actual compiled regex and then returning it.

```
    def to_python(self, value):
        """
        Handles the following cases:
        1. If the value is already the proper type (a regex), return it.
        2. If the value is a string, compile and return the regex.

        Raises: A ValidationError if the regex cannot be compiled.
        """
        if isinstance(value, type(re.compile(''))):
            return value
        else:
            if value is None and self.null:
                return None
            else:
                try:
                    return self.get_compiled_regex(value)
                except:
                    raise ValidationError('Invalid regex {0}'.format(value))
```

I wonder if there isn't a way to make a custom validator that ensures the string is a legitimate regex (for case 1) during validation and then changing the value of the field during the post_init phase of the model lifecycle.  It seems like you'd also need a corresponding thing in the pre_save to turn the regex back into a pattern string before it gets saved also.

**Comment by MikeSandford** [comment]
While I wasn't able to figure out everything related to the workaround I've described, creating a custom validator to ensure that the string is a compilable regex wasn't too hard:

```
def validate_real_regex(value):
    try:
        regex = re.compile(value)
    except re.error as e:
        raise ValidationError(
            _('%(value)s is not a valid regex: %(error)'),
            params={'value': value, 'error': str(e)},
        )
```

Then using that is "as simple as" this: 
`url_re = models.CharField(max_length=200, blank=True, null=True, validators=[validate_real_regex])`

**Comment by Benbb96** [comment]
Thank you @MikeSandford for this example, I use it now instead of the django-regex-validator.

I just wanted to say that I had to add an '_s_' after **%(error)** to make the exception work as wanted.
Otherwise, it was a **ValueError** (_incomplete format_).

**Comment by Fandekasp** [comment]
@wesleykendall @wesokes Can you explain why this bug from 2015 and still ongoing in 2018 hasn't been addressed yet ?

**Comment by wesokes** [comment]
@Bartvds @MikeSandford @Benbb96 @Fandekasp 
This is fixed in version 1.2.0. If you still have any issues let me know. Here is the pr containing the fix (and other fixes)

https://github.com/ambitioninc/django-regex-field/pull/30

Django converts the field from a string to a regex before running validators so it is trying to run a max length check on a regex object. It looks like django adds a default max length validator internally even if you don't specify one. The django admin form must call .clean on the model, which tries to run validators. My fix for this was to override the run_validators method to call the parent method with the string representation of the regex.

**Comment by MikeSandford** [comment]
Thanks for fixing!  Super stoked to see that it got squared away.
